File: code/read-lat-long/gps_serial_port.m
clear; clc;
s = serialport('COM7', 57600);
i = 0; % packet number
n = 30; % number of  packets to receive
coordinate = single(zeros(1,2)); % (latitude, longitude)
while ( i <= n )
    if ( read(s, 1, 'uint8') == 'h' )
        if (i == 0)
            flush(s);
            fprintf('The serial port buffer is flushed. Data transfer begins now.\n')
            fprintf('Total of %i packets will be received.\n', n);
        else
            coordinate(1) = read(s, 1, 'single');
            coordinate(2) = read(s, 1, 'single');
            fprintf('NumOfBytesAvailable = %i    Packet = %i    Latitude = %.5f    Longitude = %.5f\n', ...
                s.NumBytesAvailable, i, coordinate(1), coordinate(2));
        end
        i = i+1; % next packet
    end
end
delete(s);

File: code/read-lat-long/gps_serial_port.py
import numpy as np
import serial
ser = serial.Serial()
ser.baudrate = 57600
ser.port = 'COM7'
ser.open()
if (ser.is_open):
    print('Serial port %s with %i baudrate is open.' %(ser.port, ser.baudrate))
coordinate = np.zeros(2, dtype='float32') # latitude, longitude
i = 0 # packet number
n = 30 # how many packets to receive
while i <= n:
    if ser.read() == b'\x68': # corresponds to 'h'
        if i == 0:
            ser.flush()
            print('The serial port buffer is flushed. Data transfer begins now.')
            print('Total of %i packets will be received.' %n)
        else:
            coordinate[0] = np.frombuffer(ser.read(4), dtype='float32')
            coordinate[1] = np.frombuffer(ser.read(4), dtype='float32')
            print('Packet = %i    Latitude = %.5f    Longitude = %.5f' %(i, coordinate[0], coordinate[1]))
        i = i + 1  
ser.close()
if (not ser.is_open):
    print('Serial port is terminated.')